[OOP - 로또] 1단계 미션 제출합니다.

bigpel66 opened the pull request:
1단계를 진행하면서 3가지 부분이 흐지부지 된다는 느낌을 받았는데요.

1) 처음 설계할 때는 원시 값을 최대한 감싸서 객체 운용으로 진행하려고 했으나, 시간이 자날수록 이 부분이 흐지부지 되는 것을 느꼈던 거 같습니다.

2) 둘째로는 메인 로직의 흐름을 명확히 인지하지 않다보니, (지금은 수정했지만) LottoGame의 인자를 구성할 때 한 번에 WinningLotto를 만들었습니다. 이 때문에 제출 직전에 급하게 어플리케이션을 완성하려다보니 처음처럼 집중해서 객체를 만들거나 메서드를 작성하는 것이 어려웠습니다.

3) 추가로 테스트하기 어려운 부분들을 주입으로 해결하려고 객체 상위 단으로 올렸었는데, 현재 코드를 보시면 아시다시피 Lottos를 구성하는 것이 List<Lotto>가 아니라 Strategy를 넘기기 때문에 테스트가 불완전하다는 느낌도 받았습니다.

미리 리뷰 감사합니다.

htrucci commented on src/main/java/lotto/view/LottoOutputView.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.Arrays;
+
+public class LottoOutputView {
+    public static final String PURCHASE_COUNT_MESSAGE = "%d개를 구매했습니다.%n";
+    public static final String WINNING_RESULT_MESSAGE = "\n당첨 통계\n---------";
+    public static final String PROFIT_RATIO_MESSAGE = "총 수익률은 %s%%입니다.";
+    public static final String RANK_RESULT_MESSAGE = "%s (%d원) - %d개";
```
💯

htrucci commented on src/test/java/lotto/LottoTest.java:
```diff
@@ -0,0 +1,92 @@
+        assertThatThrownBy(() -> new Lotto("1, a, 2, 3, 4, 5")).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("Lotto 번호가 중복되었을 경우 예외 발생")
+    void lottoDuplicateTest() {
+        assertThatThrownBy(() -> new Lotto("1, 1, 2, 3, 4, 5")).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("Lotto 번호가 6개가 아닌 경우 예외 발생")
+    void lottoCountTest() {
+        assertThatThrownBy(() -> new Lotto("1, 2, 3, 4, 5, 6, 7")).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("Lotto 번호가 중복되고 번호가 6개가 아닌 경우 예외 발생")
+    void lottoDuplicateCountTest() {
+        assertThatThrownBy(() -> new Lotto("1, 1, 2, 3, 4, 5, 6")).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("Lotto 번호를 자동으로 생성했을 때 범위가 올바른지 테스트")
+    void randomNumberRangeTest() {
+        List<String> lottoNumbers = Lotto.autoGenerate();
+        assertThat(lottoNumbers).allMatch((lottoNumber) -> (1 <= Integer.parseInt(lottoNumber) && Integer.parseInt(lottoNumber) <= 45));
+    }
```
테스트코드를 아주 상세하게 잘 짜셨네요!
다만, 전체적인 회귀 테스트 수준으로 진행되고 있어 작은 기능 단위로 테스트가 되면 더 좋지 않을까 싶습니다.
(checkCount 라든가 checkDuplicate를 직접 테스트하는 추가)

htrucci reviewed (approved):
안녕하세요, 로또 미션 리뷰를 하게 된 데빈입니다.
발생가능한 예외에 대한 처리도 미리 해주시고 객체지향적으로 클래스 설계를 아주 잘 하신 것 같습니다. 👏 
특별히 눈에 띄는 문제가 없어서 머지를 하도록 하고 몇가지 리뷰를 남겨드렸으니 
괜찮은 부분은 다음 스텝에서 반영을 하면 좋을 것 같아요 👍

jypark5826 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+        this.description = description;
+    }
+
+    public static Rank createRank(int matchedCount, boolean hasBonus) {
+        if (matchedCount == Lotto.COUNT_OF_NUMBERS) {
+            return FIRST;
+        }
+        if (matchedCount == Lotto.COUNT_OF_NUMBERS - 1 && hasBonus) {
+            return SECOND;
+        }
+        if (matchedCount == Lotto.COUNT_OF_NUMBERS - 1) {
+            return THIRD;
+        }
+        if (matchedCount == Lotto.COUNT_OF_NUMBERS - 2) {
+            return FOURTH;
+        }
+        if (matchedCount == Lotto.COUNT_OF_NUMBERS - 3) {
+            return FIFTH;
+        }
+        return NOTHING;
+    }
```
질문있습니다!
switch case를 indent depth 1단계로 하려면 어떻게 해야하나요?

owenoweny committed:
feat: 로또 클래스 작성

- 랜덤 로또 발행 로직, 단위 테스트 구현

owenoweny committed:
docs: 기능 목록 작성

owenoweny committed:
feat: 로또 발행 기능 구현

owenoweny committed:
feat: 구입 금액에 따른 로또 발행 로직 구현

- 도메인 클래스 패키지 이동
- Lottos.java, LottoMachine.java 구현

owenoweny committed:
feat: 로또 당첨 결과 계산 기능 구현

owenoweny committed:
feat: 수익률 계산 기능 구현

Commit:
docs: 기능 목록 구체화

Commit:
feat: 로또 숫자들을 자동으로 생성하는 전략 작성

Commit:
feat: 랜덤한 수로 구성된 1 & n장의 Lotto 발행을 구현
- AutoLottoNumberStrategy에서의 상수를 별도 enum으로 분리
- int 값을 Wrapping한 LottoNumber와 테스트 작성
- List<LottoNumber>를 Wrapping한 Lotto와 테스트 작성
- Lotto와 LottoNumber를 갖고 있는 WinningLotto 및 테스트 작성
- Lotto 다발인 Lottos와 테스트 작성

Commit:
feat: LottoMachine을 구현
- 구입 금액 투입 시 살 수 있는 로또 다발 수를 결정
- 로또 가격을 LottoCondition에 추가

Commit:
feat: 당첨 판별 로직 작성
- Rank 객체 구현과 관련 상수들을 재정의 (상금 및 순위 판별 로직)
- 순위를 결정지을 수 있도록 Lotto 객체 내에서 matchCount와 matchBonus 여부를 반환하도록 구현

Commit:
feat: LottoGame을 작성
- Game을 시작함에 따라 Lottos와 WinningLotto를 구성
- start 시 복권을 긁기 시작하여 상금을 계산
- 게임을 마쳤을 때 수익률을 반환

Commit:
feat: Application 작성
- ConsoleView 작성에 따라 메시지를 출력하면서 입력 값을 구성
- LottoGame의 경우 입출력의 순서에 따라 구성이 달라지면서 생성자와 start의 인자를 변경
- 위 로직에 따라 테스트 코드도 변경

Commit:
docs: 기능 구현 완료 처리

Commit:
refactor: 로또 길이가 가변이 되는 경우를 고려했음에도, 뷰에서는 고려되지 않은 부분을 수정

Commit:
refactor: LottoGame이 직접적으로 전략으로 Lottos를 생성하는 것이 아니라 주입받도록 변경
- Application 로직을 변경
- ConsoleView의 출력 문구룰 변경

Commit:
refactor: Lottos를 List<Lotto>를 주입 받아 생성하도록 변경
- 기존 Strategy를 이용하던 부분을 Machine의 책임으로 이동

shared-moon reviewed (changes requested):
종환님 안녕하세요!
1단계 미션 잘 진행 해 주셨네요 :)
코멘트를 좀 남겨 두었는데, 확인 후 리뷰 재요청 부탁드려요 ~
고생 많으셨습니다!

Commit:
refactor: Application의 main 구문 가독성 확보
- 메서드 분리

Commit:
refactor: 1차 피드백을 반영
- LottoCondition에 작성되었던 상수들을 Lotto가 갖도록 구성
- 각 생성자들이 최상단 위치, of 메서드가 생성자 다음 위치, Equals와 HashCode는 최하단 위치로 변경
- 불필요한 방어적 복사를 지양
- 로또의 당락 여부를 로또가 들고 있지 않고, 당첨 번호가 들고 있도록 수정
- 당첨 번호에서 매번 Set을 구성하던 부분을 수정하여 단일로 생성 시 갖도록 변경

Commit:
refactor: Rank 로직의 가독성을 확보

Commit:
refactor: Strategy 삭제 및 수동 입력을 고려

shared-moon reviewed (approved):
종환님 안녕하세요!
피드백 반영 잘 진행 해 주셨네요 :)
코멘트를 좀 남겨 두었는데, 다음 미션 진행 전 확인 부탁드려요
고생 많으셨습니다 ~
